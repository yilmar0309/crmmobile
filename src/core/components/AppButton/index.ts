export {AppButton, AppButtonVariant, AppButtonSizeVariant} from './AppButton';
