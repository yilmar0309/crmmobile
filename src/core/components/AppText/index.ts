export {AppText, AppTextVariant} from './AppText';
