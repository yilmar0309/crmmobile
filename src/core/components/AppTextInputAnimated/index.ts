export {AppTextInputAnimated, TextInputState} from './AppTextInputAnimated';
