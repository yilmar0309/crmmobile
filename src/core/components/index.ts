export {AppText, AppTextVariant} from './AppText';
export {AppButton, AppButtonVariant, AppButtonSizeVariant} from './AppButton';
export {AppTextInputAnimated, TextInputState} from './AppTextInputAnimated';
export {AppContainer} from './AppContainer';
export {AppSpacer} from './AppSpacer';
export {AppHeaderLeftBack} from './AppHeaderLeftBack';
export {AppShowEmptyList} from './AppShowEmptyList';
export {AppLoading} from './AppLoading';
export {AppImage} from './AppImage';
export {AppLoadingList} from './AppLoadingList';
