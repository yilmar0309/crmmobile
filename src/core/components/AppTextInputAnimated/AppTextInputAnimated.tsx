import React, {useEffect, useRef, useState} from 'react';
import {
  Animated,
  KeyboardTypeOptions,
  Pressable,
  StyleProp,
  StyleSheet,
  TextInput,
  TextInputProps,
  useColorScheme,
  ViewStyle,
} from 'react-native';
import {useController} from 'react-hook-form';
import {View} from 'react-native-ui-lib';
import {CrossFilledIcon, InfoIconOutlined} from '@/core/assets/svg';
import {colorsDark, colorsLight} from '@/core/theme';
import {AppText, AppTextVariant} from '../AppText';

export enum TextInputType {
  FLOAT = 'FLOAT',
  OUTLINE = 'OUTLINE',
}

export enum TextInputState {
  BLUR = 'blur',
  FOCUS = 'focus',
}

interface TextInputAnimatedProps {
  control?: any;
  name: string;
  required?: boolean;
  label?: string;
  error?: boolean;
  textError?: string;
  helperText?: string;
  autoCapitalize?: 'none' | 'sentences' | 'words' | 'characters' | undefined;
  autoComplete?: TextInputProps['autoComplete'];
  keyboardType?: KeyboardTypeOptions | undefined;
  editable?: boolean;
  containerStyle?: StyleProp<ViewStyle>;
  multiline?: boolean;
  isDark?: boolean;
  secureTextEntry?: boolean;
  onChange?: (text: string) => void;
  rightIcon?: React.ReactNode;
  leftIcon?: React.ReactNode;
  maxLength?: number;
  value?: string;
  type?: TextInputType;
  placeholder?: string;
  setOnFocus?: () => void;
  setOnBlur?: () => void;
  onMessage?: (
    message: (typeof TextInputState)[keyof typeof TextInputState],
  ) => void;
  testID?: string;
}

export enum inputBorderWidth {
  isFocus = 2,
  isNotFocus = 1,
}

export enum inputBorderRadius {
  float = 5,
  outline = 10,
}

export const AppTextInputAnimated = ({
  name,
  control,
  label,
  error = false,
  required,
  textError = 'Required input',
  editable = true,
  containerStyle,
  keyboardType,
  autoCapitalize,
  autoComplete,
  multiline = false,
  secureTextEntry = false,
  onChange,
  rightIcon,
  leftIcon,
  onMessage,
  maxLength,
  value,
  type = TextInputType.OUTLINE,
  placeholder,
  helperText,
  setOnFocus,
  setOnBlur,
  testID = 'app-text-input',
}: TextInputAnimatedProps) => {
  const isDarkMode = useColorScheme() === 'dark';
  const {field} = useController({
    control,
    defaultValue: '',
    name,
    rules: {
      required,
    },
  });
  // const {shouldHandleKeyboardEvents} = useBottomSheetInternal();
  const inputValue = value ?? field.value;
  const [inputHeight, setHeight] = useState(0);
  const [labelWidth, setWidth] = useState(0);
  const [labelBackground, setLabelBackground] = useState<string>('transparent');
  const animation = useRef(new Animated.Value(0)).current;
  const translateY = animation.interpolate({
    inputRange: [0, 1],
    outputRange: [0, -inputHeight / (type === TextInputType.FLOAT ? 2 : 5)],
  });
  const translateX = animation.interpolate({
    inputRange: [0, 1],
    outputRange: [0, -labelWidth / 8],
  });
  const scale = animation.interpolate({
    inputRange: [0, 1],
    outputRange: [1, 0.8],
  });
  const [isFocus, setFocus] = useState<boolean>(false);

  const onFocus = () => {
    animate(1);
    setLabelBackground(
      isDarkMode
        ? colorsDark.BACKGROUND_SCREEN_COLOR
        : colorsLight.BACKGROUND_SCREEN_COLOR,
    );
    setFocus(true);
    onMessage && onMessage(TextInputState.FOCUS);
    if (setOnFocus) {
      setOnFocus();
    }
  };
  const onBlur = () => {
    if (!inputValue && !placeholder) {
      animate(0);
      setLabelBackground('transparent');
    }
    setFocus(false);
    onMessage && onMessage(TextInputState.BLUR);
    if (setOnBlur) {
      setOnBlur();
    }
  };
  const animate = (val: number) => {
    Animated.spring(animation, {
      toValue: val,
      bounciness: 0,
      useNativeDriver: true,
    }).start();
  };

  useEffect(() => {
    if (inputValue) {
      onFocus();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [inputValue]);

  useEffect(() => {
    if (placeholder) {
      animate(1);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [placeholder]);

  return (
    <View style={styles.container} testID={testID}>
      <View
        paddingH-14
        style={[
          styles.inputContainer,
          containerStyle,
          {
            borderColor: isFocus
              ? colorsLight.PRIMARY_COLOR
              : isDarkMode
              ? colorsLight.BACKGROUND_SCREEN_COLOR
              : colorsDark.BACKGROUND_SCREEN_COLOR,
            borderWidth: isFocus
              ? inputBorderWidth.isFocus
              : inputBorderWidth.isNotFocus,
            borderRadius:
              type === TextInputType.OUTLINE
                ? inputBorderRadius.outline
                : inputBorderRadius.float,
          },
          type === TextInputType.OUTLINE && {
            backgroundColor: isDarkMode
              ? colorsDark.BACKGROUND_SCREEN_COLOR
              : colorsLight.BACKGROUND_SCREEN_COLOR,
          },
          error && {borderColor: colorsLight.ERROR_COLOR},
        ]}
        onLayout={(e: any) =>
          !inputHeight && setHeight(e.nativeEvent.layout.height)
        }>
        <View style={{height: inputHeight, ...styles.labelContainer}}>
          <Animated.Text
            style={[
              styles.label,
              isDarkMode ? styles.labelDark : styles.labelLight,
              {transform: [{translateY}, {translateX}, {scale}]},
              {backgroundColor: labelBackground},
              {
                color: isFocus
                  ? colorsLight.PRIMARY_COLOR
                  : isDarkMode
                  ? colorsDark.BUTTON_TEXT_COLOR
                  : colorsLight.BUTTON_TEXT_COLOR,
              },
              error && {color: colorsLight.ERROR_COLOR},
            ]}
            onTextLayout={e =>
              !labelWidth && setWidth(e.nativeEvent.lines[0]?.width || 0)
            }>
            {label}
          </Animated.Text>
        </View>
        <View style={styles.iconStyles}>
          {leftIcon ?? null}
          <TextInput
            testID="app-text-input"
            style={[
              styles.input,
              type === TextInputType.FLOAT
                ? styles.floatInput
                : styles.outlineInput,
              isDarkMode ? styles.inputDark : styles.inputLight,
              multiline && styles.inputArea,
            ]}
            autoComplete={autoComplete}
            value={inputValue}
            editable={editable}
            keyboardType={keyboardType}
            autoCapitalize={autoCapitalize}
            secureTextEntry={secureTextEntry}
            multiline={multiline}
            placeholder={placeholder}
            onBlur={onBlur}
            onFocus={onFocus}
            onChangeText={onChange || field.onChange}
            maxLength={maxLength}
            caretHidden={false}
          />
          {rightIcon ?? null}
          {editable && field.value && isFocus && (
            <Pressable
              style={styles.closeIcon}
              onPress={e => {
                e.stopPropagation();
                onChange || field.onChange('');
              }}>
              <CrossFilledIcon
                color={
                  isDarkMode
                    ? colorsDark.PRIMARY_COLOR
                    : colorsLight.PRIMARY_COLOR
                }
              />
            </Pressable>
          )}
        </View>
      </View>
      {error && (
        <>
          <View row marginT-4 paddingH-16 centerV style={styles.errorContainer}>
            <InfoIconOutlined
              color={
                isDarkMode ? colorsDark.ERROR_COLOR : colorsLight.ERROR_COLOR
              }
              width={12}
              height={12}
            />
            <AppText
              variant={AppTextVariant.body1}
              color={
                isDarkMode ? colorsDark.ERROR_COLOR : colorsLight.ERROR_COLOR
              }>
              {textError ?? ''}
            </AppText>
          </View>
        </>
      )}
      {helperText && !error && (
        <>
          <View row marginT-4 paddingH-16 style={styles.errorContainer}>
            <AppText variant={AppTextVariant.body1} style={styles.helpText}>
              {helperText}
            </AppText>
          </View>
        </>
      )}
      {maxLength && !error && (
        <AppText
          variant={AppTextVariant.body1}
          style={isDarkMode ? styles.maxLengthDark : styles.maxLengthLight}>
          {inputValue.length}/{maxLength}
        </AppText>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: '100%',
  },
  inputContainer: {
    borderColor: colorsLight.OUTLINE,
    borderWidth: inputBorderWidth.isNotFocus,
    width: '100%',
  },
  input: {
    flex: 1,
    fontWeight: '400',
    lineHeight: 24,
    letterSpacing: 0.15,
  },
  floatInput: {
    fontSize: 18,
    height: 48,
  },
  outlineInput: {
    marginTop: 2,
    marginBottom: -8,
    fontSize: 16,
    height: 64,
  },
  inputDark: {
    color: colorsDark.CONTENT_PRIMARY,
  },
  inputLight: {
    color: colorsLight.CONTENT_SECONDARY,
  },
  inputArea: {
    height: 100,
    textAlignVertical: 'top',
  },
  labelContainer: {
    position: 'absolute',
    justifyContent: 'center',
  },
  label: {
    fontSize: 16,
    position: 'absolute',
    fontFamily: 'Poppins',
    marginLeft: 16,
    fontWeight: '400',
    lineHeight: 24,
    letterSpacing: 0.15,
  },
  labelDark: {
    color: colorsDark.CONTENT_SECONDARY,
  },
  labelLight: {
    color: colorsLight.CONTENT_SECONDARY,
  },
  maxLengthDark: {
    color: colorsDark.CONTENT_SECONDARY,
    marginTop: 8,
    alignSelf: 'flex-end',
    paddingHorizontal: 16,
  },
  maxLengthLight: {
    color: colorsLight.CONTENT_SECONDARY,
    marginTop: 8,
    alignSelf: 'flex-end',
    paddingHorizontal: 16,
  },
  helpText: {
    color: colorsLight.CONTENT_SECONDARY,
  },
  iconStyles: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  closeIcon: {
    marginLeft: 8,
  },
  errorContainer: {
    gap: 8,
  },
});
