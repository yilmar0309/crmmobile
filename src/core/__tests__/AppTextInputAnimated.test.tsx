import React from 'react';
import {Animated, Pressable} from 'react-native';
import {fireEvent, render} from '@testing-library/react-native';
import {AppTextInputAnimated} from '../components/AppTextInputAnimated';
import {TextInputType} from '../components/AppTextInputAnimated/AppTextInputAnimated';
const mockedAppearance = require('react-native/Libraries/Utilities/Appearance');

describe('AppTextInputAnimated', () => {
  const baseProps = {
    name: 'test-input',
    label: 'Test Input',
    placeholder: 'Type here',
    control: {},
  };

  beforeEach(() => {
    mockedAppearance.setColorScheme('dark');
  });

  afterEach(() => {
    mockedAppearance.setColorScheme('light');
  });

  it('should render input with placeholder', () => {
    const {getByPlaceholderText, toJSON} = render(
      <AppTextInputAnimated {...baseProps} testID="app-text-input-1" />,
    );
    expect(getByPlaceholderText('Type here')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('should display label and handle focus and blur events', () => {
    mockedAppearance.setColorScheme('dark');
    const onFocusMock = jest.fn();
    const onBlurMock = jest.fn();
    const {getByTestId} = render(
      <AppTextInputAnimated
        testID="app-text-input-2"
        {...baseProps}
        setOnFocus={onFocusMock}
        setOnBlur={onBlurMock}
      />,
    );

    const inputField = getByTestId('app-text-input');
    fireEvent(inputField, 'focus');
    expect(onFocusMock).toHaveBeenCalledTimes(1);
    fireEvent(inputField, 'blur');
    expect(onBlurMock).toHaveBeenCalledTimes(1);
  });

  it('should display label and handle focus and blur events if isDark is light', () => {
    mockedAppearance.setColorScheme('light');
    const onFocusMock = jest.fn();
    const onBlurMock = jest.fn();
    const {getByTestId} = render(
      <AppTextInputAnimated
        testID="app-text-input-2"
        {...baseProps}
        setOnFocus={onFocusMock}
        setOnBlur={onBlurMock}
      />,
    );

    const inputField = getByTestId('app-text-input');
    fireEvent(inputField, 'focus');
    expect(onFocusMock).toHaveBeenCalledTimes(1);
    fireEvent(inputField, 'blur');
    expect(onBlurMock).toHaveBeenCalledTimes(1);
  });

  it('should display error message when error is true', () => {
    mockedAppearance.setColorScheme('light');
    const errorMessage = 'Error occurred';
    const {getByText} = render(
      <AppTextInputAnimated
        testID="app-text-input-3"
        {...baseProps}
        error={true}
        textError={errorMessage}
      />,
    );
    expect(getByText(errorMessage)).toBeTruthy();
  });

  it('should handle text change correctly', () => {
    const onChangeMock = jest.fn();
    const {getByTestId} = render(
      <AppTextInputAnimated
        {...baseProps}
        onChange={onChangeMock}
        testID="app-text-input-4"
      />,
    );
    const inputField = getByTestId('app-text-input');
    fireEvent(inputField, 'onChangeText', 'New text');

    expect(onChangeMock).toHaveBeenCalledWith('New text');
  });

  it('should display and interact with right icon when provided', () => {
    const onIconPress = jest.fn();
    const RightIcon = () => (
      <Pressable testID="right-icon" onPress={onIconPress} />
    );
    const {getByTestId} = render(
      <AppTextInputAnimated
        {...baseProps}
        rightIcon={<RightIcon />}
        testID="app-text-input-5"
        type={TextInputType.FLOAT}
      />,
    );

    fireEvent.press(getByTestId('right-icon'));
    expect(onIconPress).toHaveBeenCalled();
  });

  it('should animate to initial state and set label background to transparent on blur when input is empty', () => {
    const animateSpy = jest.spyOn(Animated, 'spring');
    const {getByTestId} = render(
      <AppTextInputAnimated {...baseProps} testID="app-text-input-6" />,
    );

    const inputField = getByTestId('app-text-input-6');
    fireEvent.changeText(inputField, '');
    fireEvent(inputField, 'blur');

    expect(animateSpy).toHaveBeenCalledWith(expect.anything(), {
      toValue: 1,
      bounciness: 0,
      useNativeDriver: true,
    });
  });
});
