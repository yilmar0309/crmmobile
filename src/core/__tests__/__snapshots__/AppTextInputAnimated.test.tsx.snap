// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppTextInputAnimated should render input with placeholder 1`] = `
<View
  style={
    [
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      {
        "flex": 1,
        "width": "100%",
      },
    ]
  }
  testID="app-text-input-1"
>
  <View
    onLayout={[Function]}
    paddingH-14={true}
    style={
      [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        {
          "paddingHorizontal": 14,
        },
        {},
        {},
        [
          {
            "borderColor": "#BEC9C6",
            "borderWidth": 1,
            "width": "100%",
          },
          undefined,
          {
            "borderColor": "#FFFFFF",
            "borderRadius": 10,
            "borderWidth": 1,
          },
          {
            "backgroundColor": "#2C3E50",
          },
          false,
        ],
      ]
    }
  >
    <View
      style={
        [
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "height": 0,
            "justifyContent": "center",
            "position": "absolute",
          },
        ]
      }
    >
      <Text
        collapsable={false}
        onTextLayout={[Function]}
        style={
          {
            "backgroundColor": "transparent",
            "color": "#FFFFFF",
            "fontFamily": "Poppins",
            "fontSize": 16,
            "fontWeight": "400",
            "letterSpacing": 0.15,
            "lineHeight": 24,
            "marginLeft": 16,
            "position": "absolute",
            "transform": [
              {
                "translateY": 0,
              },
              {
                "translateX": 0,
              },
              {
                "scale": 1,
              },
            ],
          }
        }
      >
        Test Input
      </Text>
    </View>
    <View
      style={
        [
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <TextInput
        caretHidden={false}
        editable={true}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[MockFunction]}
        onFocus={[Function]}
        placeholder="Type here"
        secureTextEntry={false}
        style={
          [
            {
              "flex": 1,
              "fontWeight": "400",
              "letterSpacing": 0.15,
              "lineHeight": 24,
            },
            {
              "fontSize": 16,
              "height": 64,
              "marginBottom": -8,
              "marginTop": 2,
            },
            {
              "color": "#E1E2E1",
            },
            false,
          ]
        }
        testID="app-text-input"
        value=""
      />
    </View>
  </View>
</View>
`;
